pkg/events/events.go

```diff
@@ -164,6 +196,29 @@ func New(ctx context.Context, name string, identifiers CombinedIdentifiers, data
 	if identifiers != nil {
 		evt.innerEvent.Identifiers = identifiers.CombinedIdentifiers().GetEntityIdentifiers()
 	}
+	authentication := &ttnpb.Event_Authentication{}
+	if p, ok := peer.FromContext(ctx); ok && p.Addr != nil && p.Addr.String() != "pipe" {
+		if host, _, err := net.SplitHostPort(p.Addr.String()); err == nil {
+			authentication.RemoteIP = host
+		}
+	}
+	md := rpcmetadata.FromIncomingContext(ctx)
+	if md.AuthType != "" {
+		authentication.Type = md.AuthType
+	}
+	if md.AuthValue != "" {
+		if tokenType, tokenID, _, err := auth.SplitToken(md.AuthValue); err == nil {
+			authentication.TokenType = tokenType.String()
+			authentication.TokenID = tokenID
+		}
+	}
+	if md.XForwardedFor != "" {
+		xff := strings.Split(md.XForwardedFor, ",")
+		authentication.RemoteIP = strings.Trim(xff[0], " ")
+	}
+	if !authentication.Equal(&ttnpb.Event_Authentication{}) {
```

Comment:
Do we really need to allocate a new object on the heap to check if this is empty (note that you're actually constructing it here, so you definitely can determine if it will be empty without doing this)?
Sure, this will not be the performance bottleneck in production, but quite a bad code smell IMO.